torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py

```diff
@@ -0,0 +1,165 @@
+                "by DistributedDataParall in the backward pass. Therefore, only "
+                "DistributedDataParallel should be used for this case."
+            )
+        ovall_group : dist.ProcessGroup = (
+            process_group if process_group is not None else dist.group.WORLD
+        )
+        overall_group_size = dist.get_world_size(group=ovall_group)
+        if list(period_group_size_dict.values())[-1] != overall_group_size:
+            raise ValueError(
+                "The last value in arg ``period_process_group_dict`` "
+                "must be equal to the size of arg ``process_group``.")
+
+        self.period_process_group_dict = OrderedDict()
+        logger.info("Model averaging hierarchy:")
+        for period, group_size in period_group_size_dict.items():
+            logger.info(
+                f"\tEach group that has {group_size} processes average parameters every {period} iterations, "
+                "if no higher-level averaging.")
+            if group_size != overall_group_size:
+                self.period_process_group_dict[period], subgroups = dist.new_subgroups(
+                    group_size=group_size, group=ovall_group)
+            else:
+                self.period_process_group_dict[period] = ovall_group
+
+        if warmup_steps < 0:
+            raise ValueError("Arg ``warmup_steps`` must be a non-negative number.")
+        self.warmup_steps = warmup_steps
+        self.step = 0
+        self.group = ovall_group  # The process group used for averaging parameters.
+
+    def _is_process_group_found(self):
+        """
+        Returns whether ``step`` can be divided by a period in the keys of ``period_process_group_dict``.
+        If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
+        then sets set ``_group`` as the process group corresponding to the largest period,
+        since this process group will be used for averaging parameters at this ``step``.
+        """
+        for period in reversed(self._periods):
+            if self.step % period == 0:
+                self.group = self.period_process_group_dict[period]
+                return True
+        return False
+
+    def average_parameters(self, params):
+        r"""
+        Averages parameters if ``step`` is no less than ``warmup_steps``
+        and it can be divided by a period in the keys of ``period_process_group_dict``,
+        where ``step`` is increased by 1 at each iteration in the training loop.
+        If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
+        only the largest period is used, and the corresponding process group is used for averaging parameters.
+        """
+        if (
+            self.step >= self.warmup_steps
+            and self._is_process_group_found()
+        ):
+            utils.average_parameters(iter(params), self.group)
```

Comment:
unset `self.group` back to `None` after this so it is clear that it will need to be recomputed later?